### ISSUE TITLE
Exception: 'java.security.PublicKey java.security.KeyPair.getPublic()' on a null object reference

### ISSUE THREAD

**Issue by kaveeshkanwal** [ibauersachs/otr4j#31]
I am getting this exception while trying to create a session between 2 users

**Comment by gpolitis** [comment]
Hi @kaveeshkanwal, could you please post a stack trace?

**Comment by kaveeshkanwal** [comment]
Hi @gpolitis ,
Thanks for the response.
I am using this library in an android chatting project.

I am getting the above mentioned exception in AuthContextImpl.java line number 106
```
SignatureM m = new SignatureM((DHPublicKey) getLocalDHKeyPair()
                        .getPublic(), getRemoteDHPublicKey(),
                        getLocalLongTermKeyPair().getPublic(),  // LINE 106
                        getLocalDHKeyPairID());

```
```
public KeyPair getLocalLongTermKeyPair() throws OtrException {
        if (localLongTermKeyPair == null) {
            localLongTermKeyPair = getSession().getLocalKeyPair(); // THIS RETURNS NULL
        }
        return localLongTermKeyPair;
    }
```

While debugging i have seen that both private and public keys get generated, but the above code is giving me a NPE. 
I am still trying and debugging the code, but it would be great if you could help me regarding this issue.

Stack Trace:

> java.lang.NullPointerException: Attempt to invoke virtual method 'java.security.PublicKey java.security.KeyPair.getPublic()' on a null object reference
>                                                                  at net.java.otr4j.session.AuthContextImpl$MessageFactory.getRevealSignatureMessage(AuthContextImpl.java:106)
>                                                                  at net.java.otr4j.session.AuthContextImpl$MessageFactory.access$1200(AuthContextImpl.java:83)
>                                                                  at net.java.otr4j.session.AuthContextImpl.handleDHKeyMessage(AuthContextImpl.java:619)
>                                                                  at net.java.otr4j.session.AuthContextImpl.handleReceivingMessage(AuthContextImpl.java:419)
>                                                                  at net.java.otr4j.session.SessionImpl.transformReceiving(SessionImpl.java:330)
>                                                                  at net.java.otr4j.OtrEngineImpl.transformReceiving(OtrEngineImpl.java:67)
>                                                                  at com.mmi.mchat.fragments.ChatFragment.startSession(ChatFragment.java:208)
>                                                                  at com.mmi.mchat.fragments.ChatFragment.onCreateView(ChatFragment.java:149)


This entire situation occurs when i am trying to start a secure session between two users:
Below is the code that i am using.
```
private void startSession() {
        host = new DummyOtrEngineHost(new OtrPolicyImpl(OtrPolicy.ALLOW_V2
                | OtrPolicy.ERROR_START_AKE));

        usAlice = new OtrEngineImpl(host);
        usBob = new OtrEngineImpl(host);

        try {
            usAlice.startSession(aliceSessionID);
            //usBob.startSession(bobSessionID);

            // Bob receives query, sends D-H commit.
            usBob.transformReceiving(bobSessionID, host.lastInjectedMessage);

            // Alice received D-H Commit, sends D-H key.
            usAlice.transformReceiving(aliceSessionID, host.lastInjectedMessage);


            // Bob receives D-H Key, sends reveal signature. 
           //_THIS is where the entire process starts to give NPE_.
            usBob.transformReceiving(bobSessionID, host.lastInjectedMessage);


            // Alice receives Reveal Signature, sends signature and goes secure.
            usAlice.transformReceiving(aliceSessionID, host.lastInjectedMessage);

            // Bobs receives Signature, goes secure.
            usBob.transformReceiving(bobSessionID, host.lastInjectedMessage);

            if (usBob.getSessionStatus(bobSessionID) != SessionStatus.ENCRYPTED
                    || usAlice.getSessionStatus(aliceSessionID) != SessionStatus.ENCRYPTED)
                //fail("Could not establish a secure session.");
                Log.e(TAG, "Could not establish a secure session.");

        } catch (OtrException e) {
            e.printStackTrace();
        }
    }
```

**Comment by cobratbq** [comment]
Hi @kaveeshkanwal, this exception happens because you do not provide your own implementation of EngineHost.

To clarify, there are multiple types of keys. otr4j manages the session keys that are used while you're setting up an OTR session. However, there is also a long-term keypair that the chat client uses to authenticate, i.e. show that it is the same person as all the other times you chatted. For this, a long-term keypair is used. otr4j does not save/generate this keypair itself, but instead asks for it at the OtrEngineHost implementation. (This OtrEngineHost represents the chat software in which otr4j is embedded.)

In order to fix your problem, you would need to:
1. Implement your own OtrEngineHost according to the interface definition.
2. Upon request provide (either restore from file or generate on the fly) your long-term keypair. (Method call is 'getLocalLongTermKeyPair()'.)

You generate this keypair only once on a single computer/user account, then store if for all future uses. otr4j relies on this long-term keypair to remain the same. That's how your chat partner is able to identify that you are the same person he chatted with last time.

**Comment by kaveeshkanwal** [comment]
Hi @cobratbq ,
Thanks for the response, it worked!!

As per your direction, i wrote my own implementation of the OtrEngineHost and passed the created object to the constructor of OtrEngineImpl(...). 
For generating the keypair i had overriden the getLocalKeyPair(...) method of the OtrEngineHost interface  and simply returned the KeyPairGenerator object. This process successfully created a secure session between the two users.

Now, i will try to broaden this implementation and do it for multiple users by sending encrypted data to the Xmpp server.

**Comment by gpolitis** [comment]
Thank you @cobratbq!

**Comment by kaveeshkanwal** [comment]
Hi @gpolitis and @cobratbq ,
I am facing another issue when i establish a connection between two users (using Xmpp) and communicate between them using OTR.

**Case 1: (Working Fine)**
When is establish a connection (using Xmpp) with the same user (i,e the sender and the receiver are the same person), then i am able to encrypt and decrypt the message using OTR. 
Here the message is sent and received on the same mobile device.

**Case 2: (Not Working)**
Here i am using two mobile devices and i establish an Xmpp connection between two users (i,e the sender and the receiver are two different people) . In this case, the OTR is able to encrypt the message and i am able to send that message to my local Xmpp server. The Xmpp server now returns the same encrypted OTR message to the 2nd user, but i am not able to decrypt this OTR message that i receive from the Xmpp server i,e the method transformReceiving(...) returns me null string.

**Debugging:**
While debugging at the failure point i reach the _SessionImpl. java_ file and here there is a method named _handleDataMessage(....)_
@ line number 423 it is returning me NULL i,e "MAC verification failed, ignoring message"
```
if (!Arrays.equals(computedMAC, data.mac)) {
                    logger.finest("MAC verification failed, ignoring message");
                    return null; //line 423
                }
```
I am having a hard time finding out where am i going wrong, any help from you would be great :)
